### PR TITLE
HIVE-27854:Driver compile may causing memory leak

### PR DESCRIPTION
**Pull request by chenruotao**
HIVE-27854:Driver compile may causing memory leak

### What changes were proposed in this pull request?

I describe it in this the issuse 
https://issues.apache.org/jira/browse/HIVE-27854

### Why are the changes needed?

I describe it in this the issuse 
https://issues.apache.org/jira/browse/HIVE-27854

### Does this PR introduce _any_ user-facing change?

No

### Is the change a dependency upgrade?

No

### How was this patch tested?

I describe it in this the issuse 
https://issues.apache.org/jira/browse/HIVE-27854

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/Driver.java [hunk 1]
```diff
@@ -508,6 +508,10 @@ private void compile(String command, boolean resetTaskIds, boolean deferClose) t
     perfLogger.PerfLogBegin(CLASS_NAME, PerfLogger.COMPILE);
     lDrvState.stateLock.lock();
     try {
+      if (lDrvState.driverState == DriverState.DESTROYED || lDrvState.driverState == DriverState.CLOSED) {
+        LOG.warn("FAILED: Query command could not be compiled because driver has been cancelled, closed or destroyed.");
```
Comment (aturoczy) [inline ql/src/java/org/apache/hadoop/hive/ql/Driver.java:512]: Maybe better:
_Query compilation failed as the driver was canceled, closed, or destroyed_

**Review by aturoczy — COMMENTED** [review]
I would merge this PR with https://github.com/apache/hive/pull/4854 
Both PR are belong to the same issue. Does not need 2 PR for this. 

Btw is this resolved on master?

File: ql/src/java/org/apache/hadoop/hive/ql/Driver.java (hunk 1 above)
Comment (chenruotao) [reply ql/src/java/org/apache/hadoop/hive/ql/Driver.java:512]: But there is no driverState named canceled, i think the closed of driverStates include canceled

Comment (aturoczy) [reply ql/src/java/org/apache/hadoop/hive/ql/Driver.java:512]: ok

**Comment by chenruotao** [comment]
> I would merge this PR with #4854 Both PR are belong to the same issue. Does not need 2 PR for this.
> 
> Btw is this resolved on master?

sorry for 2 PR, because the related code in master  has a difference from brance 3.1.2, so I submit 2 PR, one for brance 3.1, one for master.  And I think there is the same problem on master, but it can not return directly, i set abort.

**Comment by aturoczy** [comment]
> > I would merge this PR with #4854 Both PR are belong to the same issue. Does not need 2 PR for this.
> > Btw is this resolved on master?
> 
> sorry for 2 PR, because the related code in master has a difference from brance 3.1.2, so I submit 2 PR, one for brance 3.1, one for master. And I think there is the same problem on master, but it can not return directly, so i set abort.

Please check on master. It would be better to have this PR on master, and we after that backport to 3.1. It could be a valuable change for all version

**Comment by chenruotao** [comment]
> > > I would merge this PR with #4854 Both PR are belong to the same issue. Does not need 2 PR for this.
> > > Btw is this resolved on master?
> > 
> > 
> > sorry for 2 PR, because the related code in master has a difference from brance 3.1.2, so I submit 2 PR, one for brance 3.1, one for master. And I think there is the same problem on master, but it can not return directly, so i set abort.
> 
> Please check on master. It would be better to have this PR on master, and we after that backport to 3.1. It could be a valuable change for all version

ok, I will do update on https://github.com/apache/hive/pull/4854 later, that is on master, please take a review

**Comment by aturoczy** [comment]
Sure! Please ping me on that PR

**Comment by chenruotao** [comment]
> @chenruotao, can we close this PR as it has a duplicate?

OK, if it doesn't need to be fixed in version 3.1

**Pull request closed without merging**